<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Convenient function to compute p-values from a gene...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for getPvalues {topGO}"><tr><td>getPvalues {topGO}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Convenient function to compute p-values from a gene expression matrix.</h2>

<h3>Description</h3>

<p>Warping function of &quot;mt.teststat&quot;, for computing p-values
of a gene expression matrix.
</p>


<h3>Usage</h3>

<pre>
   getPvalues(edata, classlabel, test = "t", alternative = c("greater", "two.sided", "less")[1],
   genesID = NULL, correction = c("none", "Bonferroni", "Holm", "Hochberg", "SidakSS", "SidakSD",
   "BH", "BY")[8]) 
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>edata</code></td>
<td>
<p>Gene expression matrix.</p>
</td></tr>
<tr valign="top"><td><code>classlabel</code></td>
<td>
<p>The phenotype of the data</p>
</td></tr>
<tr valign="top"><td><code>test</code></td>
<td>
<p>Which test statistic to use</p>
</td></tr>
<tr valign="top"><td><code>alternative</code></td>
<td>
<p>The alternative of the test statistic</p>
</td></tr>
<tr valign="top"><td><code>genesID</code></td>
<td>
<p>if a subset of genes is provided</p>
</td></tr>
<tr valign="top"><td><code>correction</code></td>
<td>
<p>Multiple testing correction procedure</p>
</td></tr>
</table>


<h3>Value</h3>

<p>An named numeric vector of p-values.
</p>


<h3>Author(s)</h3>

<p>Adrian Alexa</p>


<h3>See Also</h3>

<p><code><a href="GOTests.html">GOKSTest</a></code>,  <code><a href="groupStats-class.html">groupStats-class</a></code>,
<code><a href="getSigGroups.html">getSigGroups-methods</a></code>
</p>


<h3>Examples</h3>

<pre>

library(ALL)
data(ALL)

## discriminate B-cell from T-cell
classLabel &lt;- as.integer(sapply(ALL$BT, function(x) return(substr(x, 1, 1) == 'T')))

## Differentially expressed genes
geneList &lt;- getPvalues(exprs(ALL), classlabel = classLabel,
                       alternative = "greater", correction = "BY")

hist(geneList, 50)
</pre>

<hr /><div style="text-align: center;">[Package <em>topGO</em> version 2.34.0 <a href="00Index.html">Index</a>]</div>
</body></html>
